pageserver/src/layered_repository.rs

```diff
@@ -355,13 +355,34 @@ impl LayeredRepository {
         timelineid: ZTimelineId,
         tenantid: ZTenantId,
         data: &TimelineMetadata,
+        first_save: bool,
     ) -> Result<PathBuf> {
-        let path = conf.timeline_path(&timelineid, &tenantid).join("metadata");
-        let mut file = File::create(&path)?;
+        let timeline_path = conf.timeline_path(&timelineid, &tenantid);
+        let path = timeline_path.join("metadata");
+        // use OpenOptions to ensure file presence is consistent with first_save
+        let mut file = OpenOptions::new()
+            .write(true)
+            .truncate(!first_save)
+            .create_new(first_save)
+            .open(&path)?;
 
         info!("saving metadata {}", path.display());
 
-        file.write_all(&TimelineMetadata::ser(data)?)?;
+        let metadata_bytes = TimelineMetadata::ser(data)?;
+        // Taken from PG_CONTROL_MAX_SAFE_SIZE
+        const MAX_SAFE_SIZE: usize = 512;
+        assert!(metadata_bytes.len() <= MAX_SAFE_SIZE);
+
+        if file.write(&metadata_bytes)? != metadata_bytes.len() {
+            bail!("Could not write all the metadata bytes in a single call");
```

Comment:
How would things work after this `bail`? It it so, that there will be a file left with the `metadata` name, but incorrect contents?

Sounds a bit dangerous IMO if really the case.

Reply:
Yeah, I'm not sure about that case. I copied postgres semantics here:
https://github.com/postgres/postgres/blob/c30f54ad732ca5c8762bb68bbe0f51de9137dd72/src/common/controldata_utils.c#L212

Makes me wonder what happens if we happen to load a debugger just as this write is happening

Reply:
Maybe I'm not aware of something, but the metadata file is our idea and we control it entirely: I would've written that file "normally", entirely or not create it at all (retrying later?)

But never mind me much, not insisting you should do that, just expressing the confusion.

Reply:
Yeah, we have full control of the metadata file, it's our invention. But it's similar to Postgres' control file, so it makes sense to look there for inspiration.